src/approaches/grammar_search_invention_approach.py

```diff
@@ -828,8 +852,24 @@ def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
             # Bad case 2: transition is not optimal and sequence is a demo.
             elif on_sequence_h > best_h and is_demo:
                 score += CFG.grammar_search_on_demo_count_penalty
+            # Bad case 3: there is a "suspicious" optimal state.
+            for successor in optimal_successors:
+                # We already know that next_atoms is not suspicious.
+                if successor == frozenset(next_atoms):
```

Comment:
i guess this is like a minor efficiency to avoid calling _is_suspicious in the case where successor would trivially unify with next_atoms in the _is_suspicious loop anyway. Right?

Reply:
yeah right